File: ex78.py
class Solution:
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """

        r = [[]]

        for num in nums:
            for i in range(len(r)):
                r.append(r[i] + [num])

        return r


sol = Solution()
print(sol.subsets([1, 2, 3]))

File: code/ex78.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> res;
        vector<int> cur;
        dfs(nums, res, cur, 0);
        return res;
    }
    
private:
    void dfs(vector<int>& nums, vector<vector<int>>& res, vector<int>& cur, int s) {
        res.push_back(cur);
        if (s >= nums.size())    return;
        for (int i = s; i < nums.size(); i++) {
            cur.push_back(nums[i]);
            dfs(nums, res, cur, i + 1);
            cur.pop_back();
        }
    }
};